connectors/migrations/20241212_backfill_github_data_source_folders.ts

```diff
@@ -0,0 +1,236 @@
+import { makeScript } from "scripts/helpers";
+
+import { getGithubCodeDirectoryParentIds } from "@connectors/connectors/github/lib/hierarchy";
+import {
+  getCodeRootInternalId,
+  getDiscussionsInternalId,
+  getIssuesInternalId,
+  getMimeTypeFromGithubContentNodeType,
+  getRepositoryInternalId,
+} from "@connectors/connectors/github/lib/utils";
+import { dataSourceConfigFromConnector } from "@connectors/lib/api/data_source_config";
+import { concurrentExecutor } from "@connectors/lib/async_utils";
+import { upsertDataSourceFolder } from "@connectors/lib/data_sources";
+import {
+  GithubCodeDirectory,
+  GithubCodeFile,
+  GithubCodeRepository,
+  GithubDiscussion,
+  GithubIssue,
+} from "@connectors/lib/models/github";
+import { ConnectorModel } from "@connectors/resources/storage/models/connector_model";
+import type { DataSourceConfig } from "@connectors/types/data_source_config";
+
+async function createFolderNodes() {
+  const connectors = await ConnectorModel.findAll({
+    where: {
+      type: "github",
+    },
+  });
+
+  for (const connector of connectors) {
+    const dataSourceConfig = dataSourceConfigFromConnector(connector);
+    const repositories = await GithubCodeRepository.findAll({
+      where: {
+        connectorId: connector.id,
+      },
+    });
+
+    for (const repository of repositories) {
+      // upsert repository as folder
+      // Throws if error
+      const repoFolderId = await upsertRepositoryFolderNode(
+        repository,
+        dataSourceConfig
+      );
+
+      // Upsert Code folder if we have some (file or directory)
+      const shouldCreateCodeFolder = await repositoryContainsCode(repository);
+      if (shouldCreateCodeFolder) {
+        await upsertCodeFolderNode(
+          repository.repoId,
+          repoFolderId,
+          dataSourceConfig
+        );
+      }
+
+      const directories = await GithubCodeDirectory.findAll({
+        where: {
+          connectorId: connector.id,
+        },
+      });
+      // Upsert directories as folders, in chunks
+      await concurrentExecutor(
+        directories,
+        async (directory) => {
+          await upsertDirectoryFolderNode(
+            directory,
+            repository.id,
+            dataSourceConfig
+          );
+        },
+        { concurrency: 16 }
+      );
+    }
+
+    // Upsert issues folder if we have some
+    const repoWithIssues = await getUniqueRepositoryIdsWithIssues(connector.id);
```

Comment:
AFAICT we always sync the discussions and issues so you could run it for everyone. Here we could have an issue if a repo that has 0 issue creates its very first one (nothing user-facing ofc since we don't use the data_sources_nodes for content nodes as of now)

Reply:
Same reasoning, but fine with it 👍